### DISCUSSION TITLE
RPi compiled code

### DISCUSSION THREAD

**Discussion by Tangerino** [micropython/micropython#12189] [RP2040 / Pico]
My application runs seamlessly on macOS (for development purposes), Linux (for testing), and ESP32 (on actual devices) using the exact same codebase. For the ESP32, I compile the code into a .mpy file and utilize an OTA (Over-the-Air) system for distributing these files.

Now, I'm contemplating deploying this application on a Raspberry Pi. Can I maintain this portability? Specifically, I'm considering using mp-cross to compile and generate .mpy files tailored for the Raspberry Pi platform. Is this feasible? Directly deploying the source code isn't an option for me.

**Comment by rkompass** [comment]
Yes, course.
Note that the .mpy format is dependent on the architecture. You will have to specify another arch option for mpy-cross and keep the .mpy files separate.

**Reply by jimmo** [reply]
> Note that the .mpy format is dependent on the architecture.

@rkompass  This is _only_ true if you are using native code in your .mpy file. Otherwise the same .mpy file will run on every supported MicroPython target.

@Tangerino Assuming you are not using the `@native` decorator, you should consider the Raspberry Pi to be no different to any other Linux. i.e. using the Linux port.

> The Unix port is pretty limited for hardware/network support

The unix port has full network support, but yeah, limited hardware access. You can in theory access the UART via `/dev/tty` and a bit of stuffing around with termios, and GPIO via sysfs.

**Comment by Tangerino** [comment]
@rkompass That is the idea!!!

Micropython can compile to the following architectures

 `x86`, `x64`, `armv6`, `armv6m`, `armv7m`, `armv7em`, `armv7emsp`, `armv7emdp`, `xtensa`, `xtensawin`

So, as of today, my understanding is that we can't compile for modern RPi devices. Right?

I have some devices and using `lscpu` I get

```bash
lscpu
```
```
Architecture:                    aarch64
CPU op-mode(s):                  32-bit, 64-bit
Model:                           4
Model name:                      Cortex-A53
```
and
```
Architecture:          armv7l
Byte Order:            Little Endian
CPU(s):                4
Model:                 4
Model name:            ARMv7 Processor rev 4 (v7l)
```

**Reply by rkompass** [reply]
@Tangerino:

```py
import sys

sys_mpy = sys.implementation._mpy
print("arch: ", end='')
try:
    arch = [None, 'x86', 'x64', 'armv6', 'armv6m', 'armv7m', 'armv7em', 
        'armv7emsp', 'armv7emdp', 'xtensa', 'xtensawin'][sys_mpy >> 10]
    print(arch)
except:
    print('unknown')
```
gives me `arch: armv6m` on my modern RPi Pico W, and we can compile this.

**Reply by rkompass** [reply]
Did you mean an ordinary modern Raspberry Pi with Linux?
What does the above code snippet print when you run it in MP on such a device?

**Reply by scruss** [reply]
likely `arch: armv6` on 32-bit Raspberry Pi OS and `arch: None` on 64-bit. The latter's going to be a problem for any native code. So @Tangerino's assertion that *that we can't compile for modern RPi devices* is correct, if running a 64-bit OS.

The Unix port is pretty limited for hardware/network support, so running it on a Raspberry Pi may be a disappointment.

**Reply by rkompass** [reply]
I noticed that without using @Tangerino there is no E-mail notification, when messages are appended this way (.?!).
So if you read that only now again: Would be interesting to know what the above code snippet does print when you run it in MP on your Raspi.
I found out that the arm8, which is 64 bit, may be switched to 32 bit mode, which apparently happens in Raspbian OS and I now think that you tried two versions of Linux: An original 32 bit Raspbian where that happens, and a 64 bit version.
Did you try Micropython on both versions?
It's just curiosity on my side. But would be good to know if MP runs there, what I strongly assume.
And from my limited understanding so far: If you have an architecture which it runs on then you should also get an info of the
processor architecture, that is used and mpy-cross should also work with that architecture. I would be surprised if it were otherwise.

**Comment by Tangerino** [comment]
First step was to compile using Debian 32 bits
```
Linux pi5 4.19.66-v7+ #1253 SMP Thu Aug 15 11:49:46 BST 2019 armv7l GNU/Linux
```
```bash
lscpu
```
```
Architecture:          armv7l
Byte Order:            Little Endian
CPU(s):                4
On-line CPU(s) list:   0-3
Thread(s) per core:    1
Core(s) per socket:    4
Socket(s):             1
Model:                 4
Model name:            ARMv7 Processor rev 4 (v7l)
CPU max MHz:           1200.0000
CPU min MHz:           600.0000
BogoMIPS:              38.40
Flags:                 half thumb fastmult vfp edsp neon vfpv3 tls vfpv4 idiva idivt vfpd32 lpae evtstrm crc32
```

It seems to start, since in my folder I have a `main.py` and bunch of `.mpy` files

I stuck on the crypto library so I need tweak my import policies.

So far so good but just saw the unix port does not support UART so this does not make sense anymore :(

Carlos

**Comment by scruss** [comment]
> Would be interesting to know what the above code snippet does print when you run it in MP

That's what my output was: `arch: armv6` on 32-bit Raspberry Pi OS (armhf) and `arch: None` on 64-bit (aarch64). Running on a Raspberry Pi 4 will currently give you a 64-bit kernel, even if you're running a 32-bit userland.

Tangerino's Linux installation is quite old, running a 4.x kernel. Current versions are 6.1.21 or so. Sounds like the lack of UART on the Unix port was a showstopper.

**Reply by rkompass** [reply]
Thank you, that makes sense. MP on the Raspberry Pi seems to be a bit neglected, as there is ordinary Python.

**Comment by Tangerino** [comment]
I need to use a more capable hardware in solar plants where we have hundreds of devices to monitor and sometimes 4 or more serial ports.

**Reply by rkompass** [reply]
Do you have a lot of data incoming fast?
